diff --git a/drivers/input/touchscreen/focaltech_touch/focaltech_gesture.c b/drivers/input/touchscreen/focaltech_touch/focaltech_gesture.c
index be2a48bf1389..14731406d7dc 100644
--- a/drivers/input/touchscreen/focaltech_touch/focaltech_gesture.c
+++ b/drivers/input/touchscreen/focaltech_touch/focaltech_gesture.c
@@ -35,6 +35,9 @@
 *****************************************************************************/
 #include "focaltech_core.h"
+#ifdef CONFIG_TOUCHSCREEN_COMMON
+#include <linux/input/tp_common.h>
+#endif
 #if FTS_GESTURE_EN
 /******************************************************************************
 * Private constant and macro definitions using #define
@@ -130,6 +133,32 @@ static struct attribute_group fts_gesture_group = {
 	.attrs = fts_gesture_mode_attrs,
 };
 
+#ifdef CONFIG_TOUCHSCREEN_COMMON
+static ssize_t double_tap_show(struct kobject *kobj,
+				struct kobj_attribute *attr, char *buf)
+{
+	return sprintf(buf, "%d\n", fts_gesture_data.mode);
+}
+
+static ssize_t double_tap_store(struct kobject *kobj,
+	struct kobj_attribute *attr, const char *buf, size_t count)
+{
+	int rc, val;
+
+	rc = kstrtoint(buf, 10, &val);
+	if (rc)
+		return -EINVAL;
+
+	fts_gesture_data.mode = !!val;
+	return count;
+}
+
+static struct tp_common_ops double_tap_ops = {
+	.show = double_tap_show,
+	.store = double_tap_store
+};
+#endif
+
 /************************************************************************
 * Name: fts_gesture_show
 *  Brief:
@@ -609,6 +638,9 @@ int fts_gesture_init(struct fts_ts_data *ts_data)
 {
 	struct i2c_client *client = ts_data->client;
 	struct input_dev *input_dev = ts_data->input_dev;
+#ifdef CONFIG_TOUCHSCREEN_COMMON
+	int ret;
+#endif
 
 	FTS_FUNC_ENTER();
 	input_set_capability(input_dev, EV_KEY, KEY_POWER);
@@ -643,6 +675,14 @@ int fts_gesture_init(struct fts_ts_data *ts_data)
 	__set_bit(KEY_GESTURE_Z, input_dev->keybit);
 
 	fts_create_gesture_sysfs(client);
+
+#ifdef CONFIG_TOUCHSCREEN_COMMON
+	ret = tp_common_set_double_tap_ops(&double_tap_ops);
+	if (ret < 0)
+        	FTS_ERROR("%s: Failed to create double_tap node err=%d\n",
+			__func__, ret);
+#endif
+
 	fts_gesture_data.mode = DISABLE;
 	fts_gesture_data.active = DISABLE;
 
diff --git a/include/linux/input/tp_common.h b/include/linux/input/tp_common.h
new file mode 100644
index 000000000000..4b038a579fe3
--- /dev/null
+++ b/include/linux/input/tp_common.h
@@ -0,0 +1,15 @@
+#include <linux/kobject.h>
+
+extern bool capacitive_keys_enabled;
+extern struct kobject *touchpanel_kobj;
+
+struct tp_common_ops {
+	ssize_t (*show)(struct kobject *kobj, struct kobj_attribute *attr,
+			char *buf);
+	ssize_t (*store)(struct kobject *kobj, struct kobj_attribute *attr,
+			 const char *buf, size_t count);
+};
+
+int tp_common_set_capacitive_keys_ops(struct tp_common_ops *ops);
+int tp_common_set_double_tap_ops(struct tp_common_ops *ops);
+int tp_common_set_reversed_keys_ops(struct tp_common_ops *ops);
